### ISSUE TITLE
[Feature] SCL support

### ISSUE THREAD

**Issue by tank-trax** [thetwom/Tuner#86]
I am a microtonal guitarist. I have a classical guitar that has a moveable frets fretboard design byed Tolgahan Cogulu.

https://tolgahancogulu.com/

Your app has proven to be most helpful. I can't begin to thank you enough. I really like the 1/4 comma meantone with 15 notes. When I use 15 notes per scale the A#/Bb, D#/Eb, and G#/Ab pairs are the way to go. 

I was wondering if it would be possible to add SCL support; the ability to import SCL files? The Surge Synth Team has a library that may prove to be useful to add such a feature.

https://github.com/surge-synthesizer/tuning-library

Thanks again!!!

**Comment by thetwom** [comment]
Thanks for the suggestion. I am glad, the app has proven helpful for you. 

There was a request to allow custom scales before. The main difficulty here is the notation. While it is quite simple to provide a set of pitch values, the question is, how the notes should be displayed. For 12-tone only this is simple enough, but how to treat other scales?

Maybe you have more experience here in how other software deal with this problem. At the moment the non-12-tone scales are based on the up/down notation suggested in #19. We could of course hardcode a lot of different tone notations. But is this general enough?

**Comment by tank-trax** [comment]
I would look at two apps, Lingot and Scala.

[Lingot](https://github.com/ibancg/lingot) is a tuner which can import SCL files.

When it imports the SCL it assigns numbers to each note but also allows for editing.

This scale is 19 notes selected from 31EDO

```
! ED2-31_MOS-19_696.774194_-6_12_scala.scl
!
ED2-31 MOS-19 (696.774194 -6, 12)
 19
!
 77.41936
 116.12903
 193.54839
 270.96775
 309.67742
 387.09678
 464.51613
 503.22581
 580.64516
 619.35484
 696.77419
 774.19355
 812.90322
 890.32258
 967.74194
 1006.45161
 1083.87097
 1161.29033
 2/1
 ```
 
When imported into Lingot it uses numbers as follows:

![image](https://github.com/thetwom/Tuner/assets/41337033/117543af-b9c5-4d2f-86f7-dfc0c7dc1237)

Following the convention from [Scala](https://www.huygens-fokker.org/scala/) when applying E31 it provides the following notes:

![image](https://github.com/thetwom/Tuner/assets/41337033/f8dfd107-f77a-4d0a-9f7c-2ad2cd44e1f4)

After which the numbers can be manually edited in Lingot:

![image](https://github.com/thetwom/Tuner/assets/41337033/843deb31-8013-4b84-b8b7-48a4f41c4b0c)

Tuner with numbers

![image](https://github.com/thetwom/Tuner/assets/41337033/c61f2f07-5f76-4a80-9d54-c9f33faae43d)

Tuner modified

![image](https://github.com/thetwom/Tuner/assets/41337033/122840e7-0ac7-44e6-b1e6-731813efb42c)

Using these examples my suggestion would be to either allow for note editing or use the convention in Scala that applies a wide variety of tunings.

This [link](file:///home/studio/opt/scala-22-pc64-linux/help.htm#SET_NOTATION) provides the array that Scala provides.

Applying Scala notation could be problematic as its method offers two options for accidentals. I tend to go with next note up and next note down from the natural. In the case of what lies between E-F and B-C go with up.

![image](https://github.com/thetwom/Tuner/assets/41337033/a02a97d2-8bef-4c03-aff9-0441d967d775)

In all honesty I could live with numbers only. Notation is nice to have. That being said I think notation is doable.

Thanks for taking this into consideration.

**Comment by CatFoodEnthusiast** [comment]
maybe approximating the notes in 24edo and then allowing the user to edit the notes could do the job?

**Comment by tank-trax** [comment]
Would it be possible to make SCL import a feature limited to 12 note scales? This way the notation issue would not be a factor. 

I also think numbers for all other scales should not be an issue. Perhaps a combination of the 7 diatonic notes labelled as letters with the remaining notes as numbers.

**Comment by thetwom** [comment]
In order to support temperaments from users, there are a few adaptions needed. As said, I did a lot of refactoring lately which hopefully allows to create new UI features much easier in the future. This is mostly done now. It looks like custom temperaments are one of the top requested features, so I will try to look deeper into this soon.

**Comment by thetwom** [comment]
In the process of support custom instrument, I kept the format compatible to scl. You can test the latest version: [v7.0.0-rc3](https://github.com/thetwom/Tuner/releases/tag/v7.0.0-rc3).

To import a file, go to the temperaments dialog (e.g. pressing the center at the bottom bar), then click on the currently active temperament to change it. Now you are in the temperament manager. Here at the top right, you can click the three vertical buttons, there import and choose an scl file.

**Comment by tank-trax** [comment]
I just updated to Version 7.1.2 via Play Store.

I can see that we can add scales now, either with cents or ratios and that these scales can be 12 notes per scale but also less than or more than 12. We can also load SCL files. The only thing is I had to change the filename extension to `.txt` for it to load. That may be an Android thing.

Will test the frequencies soon but I am confident they will match up. 

Thanks for these changes.

**Comment by tank-trax** [comment]
Saving to file adds some extra info that shouldn't be in an SCL file. The [note=](url) part.

```
! version=7.1.2
!
! Ariel_1a.scl
!
Ariel 1a
12
!
 25/24    note=C#/Db
 9/8    note=D
 6/5    note=D#/Eb
 5/4    note=E
 4/3    note=F
 25/18    note=F#/Gb
 3/2    note=G
 8/5    note=G#/Ab
 5/3    note=A
 9/5    note=A#/Bb
 15/8    note=B
 2/1    note=C
```

As well saving will only save all custom files not the currently loaded one. 

As so....

```
! version=7.1.2
!
! Ariel_1a.scl
!
Ariel 1a
12
!
 25/24    note=C#/Db
 9/8    note=D
 6/5    note=D#/Eb
 5/4    note=E
 4/3    note=F
 25/18    note=F#/Gb
 3/2    note=G
 8/5    note=G#/Ab
 5/3    note=A
 9/5    note=A#/Bb
 15/8    note=B
 2/1    note=C
!
! Pythagorean_tuning_(Copy).scl
! abbreviation=Pythagorean
!
Pythagorean tuning (Copy)
12
!
 90.20    note=C#/Db
 203.90    note=D
 294.10    note=D#/Eb
 407.80    note=E
 498.00    note=F
 611.70    note=F#/Gb
 702.00    note=G
 792.20    note=G#/Ab
 905.87    note=A
 996.09    note=A#/Bb
 1109.78    note=B
 1200.00    note=C
```

**Comment by thetwom** [comment]
> Saving to file adds some extra info which isn't included in an SCL file. The note= part. 

Yes, it saves the note names, such that during a reimport, you get the same note names again. This is not a problem for scl, since it is allowed to put extra text behind the notes which shall be ignored. Note that if `note=` is not found, you still can load the file. 

> The extension cannot be changed to SCL as it has to be saved as TXT extension. If SCL is used it does not appear to save. This appears to be an Android thing.

This is a question of filters. The default text files require `txt`. One can open it up to allow other file extensions, but this quickly leads to the fact that you can choose between all file extensions. No idea, what is more preferable :-).

> I noticed that saving will only save custom files not the currently loaded one.

Yes, predefined temperaments cannot be saved.

> I then created a copy of the Pythagorean scale and then tried to save it to file and it saved the two custom scales in one file

Yes, all custom temperaments will be saved. If you want to save just one (or some specific ones), then long press on a temperament, to get into the "select"-mode. Note that if you save more than one temperament into the file, it will not fulfil the scl specifications any-more. More than one tuning makes only sense in case you want to share your temperaments with other tuner users, do backups, or other similar things ...

> Will test the frequencies soon but I am confident they will match up.

It will be definitely good to double-check everything is correct.

**Comment by thetwom** [comment]
I think, this can now be closed. If something is not working as expected or missing consider to reopen thie issue, or better, open a new one with the specific issue description.
